src/hotspot/share/utilities/growableArray.hpp

```diff
@@ -590,206 +599,148 @@ class GrowableArrayCHeapAllocator {
-      _bits(other._bits),
-      _nesting_check(other._nesting_check) {
-    assert(!on_C_heap(), "Copying of CHeap arrays not supported");
-    assert(!other.on_C_heap(), "Copying of CHeap arrays not supported");
-  }
-
-  GrowableArrayMetadata& operator=(const GrowableArrayMetadata& other) {
-    _bits = other._bits;
-    _nesting_check = other._nesting_check;
-    assert(!on_C_heap(), "Assignment of CHeap arrays not supported");
-    assert(!other.on_C_heap(), "Assignment of CHeap arrays not supported");
-    return *this;
-  }
-
-  void init_checks(const GrowableArrayBase* array) const;
-  void on_resource_area_alloc_check() const;
 #endif // ASSERT
 
-  bool on_C_heap() const        { return (_bits & 1) == 1; }
-  bool on_resource_area() const { return _bits == 0; }
-  bool on_arena() const         { return (_bits & 1) == 0 && _bits != 0; }
-
-  Arena* arena() const      { return (Arena*)_bits; }
-  MEMFLAGS memflags() const { return MEMFLAGS(_bits >> 1); }
-};
-
-// THE GrowableArray.
+// The GrowableArray internal data is allocated from either:
+//  - Resrouce area (default)
+//  - Arena
 //
-// Supports multiple allocation strategies:
-//  - Resource stack allocation: if no extra argument is provided
-//  - CHeap allocation: if memflags is provided
-//  - Arena allocation: if an arena is provided
-//
-// There are some drawbacks of using GrowableArray, that are removed in some
-// of the other implementations of GrowableArrayWithAllocator sub-classes:
-//
-// Memory overhead: The multiple allocation strategies uses extra metadata
-//  embedded in the instance.
+// Itself, it can be embedded, on stack, resource_arena or arena allocated.
 //
 // Strict allocation locations: There are rules about where the GrowableArray
 //  instance is allocated, that depends on where the data array is allocated.
 //  See: init_checks.
-
+//
+// For C-Heap allocation use GrowableArrayCHeap.
+//
+// Note, that with GrowableArray does not deallocate the allocated memory from
+// the arena / resource area, but rather just abandons it until the memory is
+// released by the arena or by the ResourceMark from the resource area.
+// Because GrowableArrays are often just abandoned rather than properly destructed,
+// we require that the element type E is trivially destructible, so that it makes no
+// difference if the destructors are called or not.
+//
+// GrowableArray is copyable, but it only creates a shallow copy. Hence, one has
+// to be careful not to duplicate the state and then diverge while sharing the
+// underlying data.
+//
 template <typename E>
 class GrowableArray : public GrowableArrayWithAllocator<E, GrowableArray<E> > {
   friend class GrowableArrayWithAllocator<E, GrowableArray<E> >;
-  friend class GrowableArrayTest;
 
-  static E* allocate(int max) {
-    return (E*)GrowableArrayResourceAllocator::allocate(max, sizeof(E));
-  }
+  // Since GrowableArray is arena / resource area allocated, it is a custom to
```

Comment:
"it is a custom to" can basically be removed

"Since Growable array is arena/resource area allocated it does not destruct its elements. Therefore, ..." is sufficient.